### PR TITLE
[Merged by Bors] - Add commmand for pruning states

### LINKED ISSUE #4481: Add logic to prune all historic states
## Description

It would be useful for several purposes to be able to drop all of the states in the freezer DB. Some uses include:

- Recover from database corruption (e.g. due to https://github.com/sigp/lighthouse/issues/3011).
- Enable historic block pruning (see https://github.com/sigp/lighthouse/pull/4158).
- Save space by running without historic states (https://github.com/sigp/lighthouse/issues/3211).
- Migrate from stable Lighthouse to `tree-states`

## Steps to resolve

I think for forwards-compatibility with tree-states it would be best to implement the deletion as several iterators over the relevant columns, which just delete everything in the column without parsing it. Basically:

```rust
for column in relevant {
  for key in db.iter_column(column) {
    to_delete.push(key);
  }
}
```

Where `relevant` consists of:

* `DBColumn::BeaconStateSummary` (where the `ColdStateSummary` is stored)
* `DBColumn::BeaconState` (where the `PartialBeaconState` is stored)
* `DBColumn::BeaconRestorePoint` (where the restore point -> state root index is stored)
* `DBColumn::BeaconStateRoots`
* `DBColumn::BeaconHistoricalRoots`
* `DBColumn::BeaconRandaoMixes`
* `DBColumn::BeaconHistoricalSummaries`

I think on `unstable` the beacon block roots should be left intact, as we expect them to exist whenever we've stored historic _blocks_ in the DB.

Atomically with those deletions we should _update_ the `state_upper_limit` in the `Anchor` to reflect the fact that those states are gone. I think setting it to `u64::MAX` could be a good way to signal a new mode of operation where "this node isn't storing any historic states". Alternatively, we set it to the next restore point slot (or snapshot slot for tree-states) which is `>= split.slot`. Either way, we need this block of logic from the tree-states branch that prevents the finalization migration from trying to write bad states:

https://github.com/sigp/lighthouse/blob/079cd67df26bd66a97f9244f540f5108c8314d9e/beacon_node/store/src/hot_cold_store.rs#L2204-L2211

I think we probably also want to solve the alignment issues for checkpoint sync first, otherwise we'll have to avoid deleting all the states when the split point is at a skipped slot. See: https://github.com/sigp/lighthouse/issues/3210#issuecomment-1588303226.

### PR DESCRIPTION
**Pull request by jimmygchen**
## Issue Addressed

Closes #4481. 

(Continuation of #4648)

## Proposed Changes

- [x] Add `lighthouse db prune-states`
- [x] Make it work
- [x] Ensure block roots are handled correctly (to be addressed in 4735)
- [x] Check perf on mainnet/Goerli/Gnosis (takes a few seconds max)
- [x] Run block root healing logic (#4875 ) at the beginning
- [x] Add some tests
- [x] Update docs
- [x] Add `--freezer` flag and other improvements to `lighthouse db inspect`

### CODE REVIEW

File: database_manager/src/lib.rs [hunk 1]
```diff
@@ -287,14 +370,17 @@ pub fn inspect_db<E: EthSpec>(
                 total += value.len();
```
Comment (jimmygchen) [inline database_manager/src/lib.rs]: Removing this line to avoid adding the value twice.

File: beacon_node/store/src/hot_cold_store.rs [hunk 2]
```diff
@@ -2254,6 +2256,92 @@ impl<E: EthSpec, Hot: ItemStore<E>, Cold: ItemStore<E>> HotColdDB<E, Hot, Cold>
...
+        genesis_state: &BeaconState<E>,
+    ) -> Result<(), Error> {
+        // Make sure there is no missing block roots before pruning
+        self.heal_freezer_block_roots()?;
+
+        // Update the anchor to use the dummy state upper limit and disable historic state storage.
+        let old_anchor = self.get_anchor_info();
+        let new_anchor = if let Some(old_anchor) = old_anchor.clone() {
+            AnchorInfo {
+                state_upper_limit: STATE_UPPER_LIMIT_NO_RETAIN,
```
Comment (jimmygchen) [inline beacon_node/store/src/hot_cold_store.rs:2278]: do we keep the `state_lower_limit` as it is? Does the value mean anything if we aren't storing historical states?

File: database_manager/src/lib.rs [hunk 3]
```diff
@@ -461,10 +576,34 @@ pub fn run<T: EthSpec>(cli_args: &ArgMatches<'_>, env: Environment<T>) -> Result
             let inspect_config = parse_inspect_config(cli_args)?;
             inspect_db(inspect_config, client_config, &context, log)
         }
-        ("prune_payloads", Some(_)) => {
+        ("prune-payloads", Some(_)) => {
             prune_payloads(client_config, &context, log).map_err(format_err)
         }
         ("prune_blobs", Some(_)) => prune_blobs(client_config, &context, log).map_err(format_err),
```
Comment (jimmygchen) [inline database_manager/src/lib.rs]:
```suggestion
        ("prune-blobs", Some(_)) => prune_blobs(client_config, &context, log).map_err(format_err),
```

File: beacon_node/store/src/hot_cold_store.rs (hunk 2 above)
Comment (jimmygchen) [inline beacon_node/store/src/hot_cold_store.rs:2278]:
```suggestion
                state_upper_limit: STATE_UPPER_LIMIT_NO_RETAIN,
                state_lower_limit: Slot::new(0),
```

Comment (michaelsproul) [reply beacon_node/store/src/hot_cold_store.rs:2278]: Good catch. It should be set to 0!

File: database_manager/src/lib.rs [hunk 4]
```diff
@@ -88,17 +89,35 @@ pub fn inspect_cli_app<'a, 'b>() -> App<'a, 'b> {
...
+    App::new("prune-payloads")
+        .alias("prune_payloads")
         .setting(clap::AppSettings::ColoredHelp)
         .about("Prune finalized execution payloads")
 }
 
 pub fn prune_blobs_app<'a, 'b>() -> App<'a, 'b> {
-    App::new("prune_blobs")
+    App::new("prune-blobs")
+        .alias("prune-blobs")
```
Comment (michaelsproul) [inline database_manager/src/lib.rs]:
```suggestion
        .alias("prune_blobs")
```

**Review by michaelsproul — APPROVED** [review]
Let's merge!

**Comment by michaelsproul** [comment]
I added and removed a backwards-incompat label because I thought we'd broken the `prune_payloads` command. But the alias exists, so it's fine.

bors r+

**Pull request closed without merging**